### ISSUE TITLE
Install NixOs from another OS: Failed to move "/boot/loader/loader.conf" into place: Operation not permitted

### ISSUE THREAD

**Issue by tobiasBora** [NixOS/nixpkgs#46098]
## Issue description

I tried to install NixOs using a Xubuntu 16.04 live USB following the instructions here : https://nixos.org/nixos/manual/index.html#sec-installing-from-other-distro. The install fails with:

 ```
$ sudo PATH="$PATH" NIX_PATH="$NIX_PATH" `which nixos-install` --root /mnt
building the configuration in /mnt/etc/nixos/configuration.nix...
[196 built, 841 copied (3363.1 MiB), 0.8 MiB DL]
installing the boot loader...
setting up /etc...
Initializing machine ID from random generator.
perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
	LANGUAGE = (unset),
	LC_ALL = (unset),
	LANG = "fr_FR.UTF-8"
    are supported and installed on your system.
perl: warning: Falling back to the standard locale ("C").
Copied "/nix/store/3hmpbbcv1db42m9g34c9g4q6qinw50x4-systemd-237/lib/systemd/boot/efi/systemd-bootx64.efi" to "/boot/EFI/systemd/systemd-bootx64.efi".
Copied "/nix/store/3hmpbbcv1db42m9g34c9g4q6qinw50x4-systemd-237/lib/systemd/boot/efi/systemd-bootx64.efi" to "/boot/EFI/BOOT/BOOTX64.EFI".
Failed to move "/boot/loader/loader.conf" into place: Operation not permitted
Traceback (most recent call last):
  File "/nix/store/4svrqgsk67dhwmlj0jqphf4yj1v0yysq-systemd-boot-builder.py", line 210, in <module>
    main()
  File "/nix/store/4svrqgsk67dhwmlj0jqphf4yj1v0yysq-systemd-boot-builder.py", line 185, in main
    subprocess.check_call(["/nix/store/3hmpbbcv1db42m9g34c9g4q6qinw50x4-systemd-237/bin/bootctl", "--path=/boot", "install"])
  File "/nix/store/i1hh6prcyfqd066jdh4zwqfs58mqhdfw-python3-3.6.6/lib/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/nix/store/3hmpbbcv1db42m9g34c9g4q6qinw50x4-systemd-237/bin/bootctl', '--path=/boot', 'install']' returned non-zero exit status 1.
```

### Steps to reproduce

 First, boot Xubuntu 16.04 liveUSB. Then I mounted the system partition on `/mnt/`, the efi partition (`/dev/sda1`) on `/mnt/boot/`, the home partition in `/mnt/home/, and then I followed all the instructions here : https://nixos.org/nixos/manual/index.html#sec-installing-from-other-distro . Note that my `configuration.nix` contains the following boot-related lines:

```
  boot.loader.systemd-boot.enable = true;
  boot.loader.efi.canTouchEfiVariables = true;
  boot.loader.grub.useOSProber = true;
```
Then, I ran the above command: `sudo PATH="$PATH" NIX_PATH="$NIX_PATH" `which nixos-install` --root /mnt` that fails with the above error.

For the complete log of all commands I typed as well as the content of the configuration files, please see: [log.txt](https://github.com/NixOS/nixpkgs/files/2352830/log.txt).

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.4.0-31-generic, Ubuntu, 16.04.1 LTS (Xenial Xerus)`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.1`
 - channels(xubuntu): `"nixpkgs-18.03.133188.8b92a4e6004"`
 - nixpkgs: `/home/xubuntu/.nix-defexpr/channels/nixpkgs`
```

Thanks for the good job!

**Comment by dezgeg** [comment]
I'm guessing this has to do with the 4.4 kernel not supporting `RENAME_NOREPLACE` which was only introduced in v4.9. Getting a strace of the failing command `/nix/store/3hmpbbcv1db42m9g34c9g4q6qinw50x4-systemd-237/bin/bootctl --path=/boot 'install` would help in confirming that. Or reproducing with a new Ubuntu.

**Comment by tobiasBora** [comment]
Indeed, I think you are right dezgeg, I get an error on the following line:

    renameat2(AT_FDCWD, "/mnt/boot/loader/.#loader.confa89a3a42cf46e36d", AT_FDCWD, "/mnt/boot/loader/loader.conf", RENAME_NOREPLACE) = -1 EINVAL (Invalid argument)

There are also a few other errors, I don't know if they are relevant:

    openat(AT_FDCWD, "/mnt/boot/loader", O_WRONLY|O_DIRECTORY|O_CLOEXEC|O_TMPFILE) = -1 EOPNOTSUPP (Operation not supported)

 Here are the last lines:

```
writev(2, [{"Copied \"/nix/store/rx76y57vl69yn"..., 142}, {"\n", 1}], 2Copied "/nix/store/rx76y57vl69ynnig325pid8jb0466298-systemd-239/lib/systemd/boot/efi/systemd-bootx64.efi" to "/mnt/boot/EFI/BOOT/BOOTX64.EFI".
) = 143
close(5)                                = 0
close(4)                                = 0
getdents(3, /* 0 entries */, 32768)     = 0
close(3)                                = 0
openat(AT_FDCWD, "/etc/machine-id", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3
read(3, "2bc0d4ded94e421c81a6a4963d213213"..., 38) = 33
read(3, "", 5)                          = 0
close(3)                                = 0
access("/mnt/boot/loader/loader.conf", F_OK) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/mnt/boot/loader", O_WRONLY|O_DIRECTORY|O_CLOEXEC|O_TMPFILE) = -1 EOPNOTSUPP (Operation not supported)
getrandom("\212\251\243$\374d>\326", 8, GRND_NONBLOCK) = 8
openat(AT_FDCWD, "/mnt/boot/loader/.#loader.confa89a3a42cf46e36d", O_WRONLY|O_CREAT|O_EXCL|O_NOCTTY|O_NOFOLLOW|O_CLOEXEC, 0640) = 3
fcntl(3, F_GETFL)                       = 0x28001 (flags O_WRONLY|O_LARGEFILE|O_NOFOLLOW)
fstat(3, {st_mode=S_IFREG|0755, st_size=0, ...}) = 0
write(3, "#timeout 3\n#console-mode keep\nde"..., 73) = 73
fsync(3)                                = 0
fstat(3, {st_mode=S_IFREG|0755, st_size=73, ...}) = 0
openat(AT_FDCWD, "/proc/self/fd/", O_RDONLY|O_DIRECTORY|O_CLOEXEC|O_PATH) = 4
readlinkat(4, "3", "/mnt/boot/loader/.#loader.confa8"..., 99) = 46
close(4)                                = 0
openat(AT_FDCWD, "/mnt/boot/loader", O_RDONLY|O_DIRECTORY|O_CLOEXEC) = 4
fsync(4)                                = 0
close(4)                                = 0
renameat2(AT_FDCWD, "/mnt/boot/loader/.#loader.confa89a3a42cf46e36d", AT_FDCWD, "/mnt/boot/loader/loader.conf", RENAME_NOREPLACE) = -1 EINVAL (Invalid argument)
newfstatat(AT_FDCWD, "/mnt/boot/loader/.#loader.confa89a3a42cf46e36d", {st_mode=S_IFREG|0755, st_size=73, ...}, AT_SYMLINK_NOFOLLOW) = 0
linkat(AT_FDCWD, "/mnt/boot/loader/.#loader.confa89a3a42cf46e36d", AT_FDCWD, "/mnt/boot/loader/loader.conf", 0) = -1 EPERM (Operation not permitted)
writev(2, [{"Failed to move \"/mnt/boot/loader"..., 81}, {"\n", 1}], 2Failed to move "/mnt/boot/loader/loader.conf" into place: Operation not permitted
) = 82
close(3)                                = 0
unlink("/mnt/boot/loader/.#loader.confa89a3a42cf46e36d") = 0
umask(022)                              = 02
exit_group(1)                           = ?
```

**Comment by tobiasBora** [comment]
You can find here the pastebin : https://paste.debian.net/1041626/.

**Comment by dezgeg** [comment]
Ok, that confirms it, thanks.

> ````
> openat(AT_FDCWD, "/mnt/boot/loader", O_WRONLY|O_DIRECTORY|O_CLOEXEC|O_TMPFILE) = -1 EOPNOTSUPP (Operation not supported)
> ````

This one is expected, as it is first tries `O_TMPFILE` if it's available. But it isn't on FAT, so it falls back to creating a random filename:

> ````
> getrandom("\212\251\243$\374d>\326", 8, GRND_NONBLOCK) = 8
> openat(AT_FDCWD, "/mnt/boot/loader/.#loader.confa89a3a42cf46e36d", O_WRONLY|O_CREAT|O_EXCL|O_NOCTTY|O_NOFOLLOW|O_CLOEXEC, 0640) = 3
> ````

and later on will try to rename it to the actual name with `renameat2(..., RENAME_NOREPLACE)`:

> ````
> renameat2(AT_FDCWD, "/mnt/boot/loader/.#loader.confa89a3a42cf46e36d", AT_FDCWD, "/mnt/boot/loader/loader.conf", RENAME_NOREPLACE) = -1 EINVAL (Invalid argument)
> ````

But since the kernel doesn't support that, it will fall back to `linkat()` (+ later `unlink()`):

> ````
> linkat(AT_FDCWD, "/mnt/boot/loader/.#loader.confa89a3a42cf46e36d", AT_FDCWD, "/mnt/boot/loader/loader.conf", 0) = -1 EPERM (Operation not permitted)
> ````

which won't work on FAT since it doesn't support hard links. And there is no fallback for that.

So it's an upstream bug which will only trigger on older kernels.

**Comment by tobiasBora** [comment]
Isn't it supposed to be easy to fix, by removing for example the RENAME_NOREPLACE flag, or just manually add a condition? Should I create a bug report on bootctl side?

**Comment by dezgeg** [comment]
Yeah please file a report at https://github.com/systemd/systemd. I presume they use RENAME_NOREPLACE there for a reason.

**Comment by lheckemann** [comment]
Should be fixed in systemd v239 and v240 (#56184)

**Comment by lheckemann** [comment]
Should be fixed now.
